google-cloud-spanner/src/main/java/com/google/cloud/spanner/Operation.java

```diff
@@ -120,43 +135,57 @@
    *
    * @return null if operation is not found otherwise the current operation.
    */
-  public Operation<R, M> waitFor(WaitForOption... options) throws SpannerException {
-    if (isDone) {
+  public Operation<R, M> waitFor(RetryOption... waitOptions) throws SpannerException {
+    if (isDone()) {
       return this;
     }
-    long timeoutMillis = Timeout.getOrDefault(options).getTimeoutMillis();
-    boolean hasTimeout = timeoutMillis != -1;
-    CheckingPeriod period = CheckingPeriod.getOrDefault(options);
-    long startMillis = clock.millisTime();
-    while (true) {
-      try {
-        com.google.longrunning.Operation proto = rpc.getOperation(name);
-        if (proto.getDone()) {
-          return Operation.<R, M>create(rpc, proto, parser);
-        }
-        long elapsed = clock.millisTime() - startMillis;
-        if (hasTimeout && elapsed >= timeoutMillis) {
-          throw SpannerExceptionFactory.newSpannerException(
-              ErrorCode.DEADLINE_EXCEEDED, "Operation did not complete in the given time");
-        }
-      } catch (SpannerException e) {
-        if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
-          return null;
-        }
-        if (!e.isRetryable()) {
-          throw e;
-        }
+    RetrySettings waitSettings =
+        RetryOption.mergeToSettings(DEFAULT_OPERATION_WAIT_SETTINGS, waitOptions);
+    try {
+      com.google.longrunning.Operation proto = RetryHelper.poll(
+          new Callable<com.google.longrunning.Operation>() {
+            @Override
+            public com.google.longrunning.Operation call() throws Exception {
+              return rpc.getOperation(name);
+            }
+          },
+          waitSettings,
+          new BasicResultRetryAlgorithm<com.google.longrunning.Operation>() {
+            @Override
+            public boolean shouldRetry(
+                Throwable prevThrowable, com.google.longrunning.Operation prevResponse) {
+              if (prevResponse != null) {
+                return !prevResponse.getDone();
+              }
+              if (prevThrowable instanceof BaseServiceException) {
+                if (!(prevThrowable instanceof SpannerException)) {
```

Comment:
why are you checking for this not to be instance of SpannerException?

Reply:
Oh, I think the idea was to "support" other retriable exceptions, which theoretically can be added in the future. 

But you are right, because of that I did it wrong, and all spanner exceptions will be retried to oblivion regardless if they are retriable or no. I've almost introduced a a time bomb bug =). Thanks!